File: src/search/binarySearch.ts
import {ICanSearch} from "./ICanSearch";
export default class BinarySearch implements ICanSearch{

    search(array: number[], target: number,low?:number,high?:number): number;
    search(array:number[],target:number,low:number,high:number): number{
        if(!high){
            high = array.length-1;
            low = 0;
        }
        return this.divide(array,target,low,high);
    }
    private divide(array:number[],target:number,low:number,high:number):number{
        if(low>=high) return -1;
        const middle = low + Math.floor((high-low) / 2);
        if(array[middle] === target){
            return middle;
        }
        if(array[middle] > target){
            return this.divide(array,target,low,middle-1);
        }
        return this.divide(array,target,middle+1,high);
    }
}

File: search/BinarySearch.java
package search;
public class BinarySearch implements ISearch{

    @Override
    public int search(int[] array, int target) {
        return this.searching(array,target,0,array.length);
    }
    public int search(int[] array, int target,int low,int high){
        return this.searching(array,target,low,high);
    }
    private int searching(int[] array, int target,int low,int high){
        if(low > high) return -1;
        int middle = low + (high-low)/2;
        if(array[middle] == target){
            return middle;
        }
        if(array[middle] > target){
            return this.searching(array,target,low,middle-1);
        }
        return this.searching(array,target,middle+1,high);
    }
}